### DISCUSSION TITLE
Type error on dynamic entity query

### DISCUSSION THREAD

**Discussion by PhentomPT** [prisma/prisma#22614] [Q&A]
### Question

On typescript i wanted to achieve the following example:
```typescript
import { prisma, Prisma } from "prisma";

async get(entity: keyof typeof Prisma.ModelName, id: number) {
    try {
      return await prisma[entity].findFirstOrThrow({ where: { id } }); // `findFirst` gives a type error
    } catch (err) {
      console.log(err);
      return { error: "unable to query" };
    }
  }
```

The type error that is given is:
`This expression is not callable. Each member of the union type '...' has signatures, but none of those signatures are compatible with each other`

I'm guessing that this is warning me because the `where` condition on every entity will have different fields but the one i want to use is present in all of them

**Comment by ludralph** [comment]
Hi @PhentomPT 👋 

Thank you for raising this question.

The error you are seeing is because TypeScript cannot fully infer types when using a dynamic key to access a model. Each model's `findFirst` method has a unique where clause signature tailored to its fields, and TypeScript cannot guarantee compatibility at compile time. One potential solution would be to assert the specific model type for `prisma[entity]` using a type assertion.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by PhentomPT** [reply]
How could i assert the type dynamically?

**Reply by daveycodez** [reply]
This is why I stuck with JavaScript for so long

**Reply by ludralph** [reply]
Hi @PhentomPT 

You could do something like:
```ts
async get(entity: keyof typeof Prisma.ModelName, id: number) {
  try {
    return await (prisma[entity] as any).findFirstOrThrow({ where: { id } });
  } catch (err) {
    console.log(err);
    return { error: "unable to query" };
  }
}
```
Alternatively, you could do:
```ts
type PrismaModelName = keyof typeof Prisma.ModelName;
const getPrismaDelegate = <N extends PrismaModelName>(name: N, prisma: PrismaClient) =>
  prisma[name.charAt(0).toLowerCase() + name.slice(1)] as PrismaModelDelegate<N>;

async function get(entity: PrismaModelName, id: number) {
  try {
    const delegate = getPrismaDelegate(entity, prisma);
    return await delegate.findFirstOrThrow({ where: { id } });
  } catch (err) {
    console.log(err);
    return { error: "unable to query" };
  }
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
